Sources/SwiftDocCUtilities/Transformers/StaticHostableTransformer.swift

```diff
@@ -0,0 +1,169 @@
+    init(dataProvider: FileSystemProvider, fileManager: FileManagerProtocol, outputURL: URL, htmlTemplate: URL, staticHostingBasePath: String?) throws {
+        self.dataProvider = dataProvider
+        self.fileManager = fileManager
+        self.outputURL = outputURL
+
+        let indexFileName = staticHostingBasePath != nil ? HTMLTemplate.templateFileName.rawValue : HTMLTemplate.indexFileName.rawValue
+        let indexFileURL = htmlTemplate.appendingPathComponent(indexFileName)
+        var indexHTML = try String(contentsOfFile: indexFileURL.path)
+
+
+        if let staticHostingBasePath = staticHostingBasePath {
+
+            var replacementString = staticHostingBasePath
+
+            // We need to ensure that the base path has a leading /
+            if !replacementString.hasPrefix("/") {
+                replacementString = "/" + replacementString
+            }
+
+            // Trailing /'s are not required so will be removed if provided.
+            if replacementString.hasSuffix("/") {
+                replacementString = String(replacementString.dropLast(1))
+            }
+
+            indexHTML = indexHTML.replacingOccurrences(of: HTMLTemplate.tag.rawValue, with: replacementString)
+        }
+        self.indexHTMLData = Data(indexHTML.utf8)
+    }
+    
+    /// Creates a static hostable version of the documention in the data folder of an archive pointed to by the `dataProvider`
+    /// - Parameters:
+    ///   - outputURL: The folder where the output will be placed
+    ///   - basePath: The path to be prefix to all href and src parameters in generated html
+    /// - Returns: An array if problems encounter during the archive.
+    func transform() throws {
+
+        let node = dataProvider.fileSystem
+        
+        // We should be starting at the data folder of a .doccarchive.
+        switch node {
+        case .directory(let dir):
+            try processDirectoryContents(directoryRoot: outputURL, relativeSubPath: "", directoryContents: dir.children)
+        case .file(_):
+            throw StaticHostableTransformerError.dataProviderDoesNotReferenceValidInput
+        }
+    }
+
+
+    /// Create a directory at the provided URL
+    ///
+    private func createDirectory(url: URL) throws {
+        if !fileManager.fileExists(atPath: url.path) {
+            try fileManager.createDirectory(at: url, withIntermediateDirectories: true, attributes: [:])
+        }
+    }
+
+    /// Processes the contents of a given directory
+    /// - Parameters:
+    ///   - root: The root output URL
+    ///   - directory: The relative path (to the root) of the directory for which then content will processed.
+    ///   - nodes: The directory contents
+    /// - Returns: An array of problems that may have occured during processing
+    private func processDirectoryContents(directoryRoot: URL, relativeSubPath: String, directoryContents: [FSNode]) throws {
```

Comment:
For my own understanding, what are these `process` methods doing, exactly? Is there a way to encapsulate this traversal behavior so it's easier to understand and more testable? For example, if these methods are used to crawl a directory and do something with each file, perhaps it'd be worth creating a type that just knows how to walk FS nodes and e.g. invoke a transformation closure or something on each file.